### PR TITLE
Expand Discord rich presence to update on section

### PR DESCRIPTION
**Pull request by vabold**
"state" and "details" in `RichPresenceManager::setGameState` would ideally be strings loaded from a BMG file, but I will leave that work for someone else to tackle in another PR.

### CODE REVIEW

File: payload/game/host_system/SystemManager.cc [hunk 1]
```diff
@@ -103,9 +106,225 @@ void RichPresenceManager::terminateConnection() {
...
+    m_finalizedStatus = true;
+    setGameState();
+}
+
+void SystemManager::LaunchTitle(u64 titleID) {
+    if (!OSIsTitleInstalled(titleID)) {
+        return;
+    }
+    __OSLaunchTitle(titleID);
+}
```
Comment (riidefi) [inline payload/game/host_system/SystemManager.cc:322]: It seems odd to have a SystemManager function inbetween RichPresenceManager functions

File: payload/game/host_system/SystemManager.cc [hunk 2]
```diff
@@ -115,11 +334,33 @@ bool RichPresenceManager::sendStatus() {
...
-    __OSLaunchTitle(titleID);
+
+    std::string_view details = "MKW-SP";
+    if (m_sceneId == RKSceneID::Menu || m_sceneId == RKSceneID::Race ||
+            m_sceneId == RKSceneID::Globe) {
+        auto *section = UI::SectionManager::Instance()->currentSection();
+        if (section && UI::Section::HasRoomServer(section->id())) {
+            details = "Server Mode";
+        }
+    }
```
Comment (riidefi) [inline payload/game/host_system/SystemManager.cc:349]: I would rather this be an argument to the function

File: payload/game/host_system/SystemManager.cc [hunk 3]
```diff
@@ -115,11 +334,33 @@ bool RichPresenceManager::sendStatus() {
...
+
+    if (!std::string_view(m_state).empty() && (!m_statusWasSent || m_state != status().state)) {
```
Comment (riidefi) [inline payload/game/host_system/SystemManager.cc:351]: I would take `state` and `timestamp` as arguments, too. Then this becomes `state != status().state || details != status().details || timestamp != status().startTimestamp`

Comment (vabold) [reply payload/game/host_system/SystemManager.cc]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/5fcc7f8188634960aefd6670d3d28127ae0dfd6c. (Details will never change without the state also changing.)

File: payload/game/host_system/SystemManager.cc [hunk 4]
```diff
@@ -103,9 +106,225 @@ void RichPresenceManager::terminateConnection() {
...
+        break;
+    case (UI::SectionId)0x78:
+    case (UI::SectionId)0x79:
+        m_state = "Disconnected from online race!";
+        break;
+    case (UI::SectionId)0x94:
+        m_state = "Configuring mission mode...";
+        break;
+    default:
+        break;
```
[the remark this answers is not in the document]
Comment (vabold) [reply payload/game/host_system/SystemManager.cc:316]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/5fcc7f8188634960aefd6670d3d28127ae0dfd6c.

File: payload/game/host_system/SystemManager.cc (hunk 2 above)
Comment (vabold) [reply payload/game/host_system/SystemManager.cc]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/5fcc7f8188634960aefd6670d3d28127ae0dfd6c.

File: payload/game/host_system/SystemManager.cc (hunk 1 above)
Comment (vabold) [reply payload/game/host_system/SystemManager.cc]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/5fcc7f8188634960aefd6670d3d28127ae0dfd6c.

File: payload/game/host_system/SystemManager.cc [hunk 5]
```diff
@@ -103,9 +106,225 @@ void RichPresenceManager::terminateConnection() {
         SP::IOSDolphin::DiscordReset();
     }
 }
+
+void RichPresenceManager::onSceneChange(RKSceneID sceneId) {
```
[the remark this answers is not in the document]
Comment (vabold) [reply payload/game/host_system/SystemManager.cc]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/5fcc7f8188634960aefd6670d3d28127ae0dfd6c.

File: payload/game/host_system/SystemManager.hh [hunk 6]
```diff
@@ -40,61 +42,40 @@ public:
...
-            sendStatus();
-        }
-    }
+    void onSceneChange(RKSceneID id);
+    void onSectionChange(UI::SectionId sectionId);
 
-    // Internal
-    PresenceData &status();
+private:
+    inline PresenceData &status();
```
[the remark this answers is not in the document]
Comment (vabold) [reply payload/game/host_system/SystemManager.hh]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/5fcc7f8188634960aefd6670d3d28127ae0dfd6c.

File: payload/game/host_system/SystemManager.cc [hunk 7]
```diff
@@ -4,6 +4,8 @@ extern "C" {
 #include <revolution.h>
 }
 
+#include "game/system/RaceManager.hh"
+
```
[the remark this answers is not in the document]
Comment (vabold) [reply payload/game/host_system/SystemManager.cc]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/5fcc7f8188634960aefd6670d3d28127ae0dfd6c.

File: payload/game/host_system/SystemManager.cc [hunk 8]
```diff
@@ -115,11 +337,16 @@ bool RichPresenceManager::sendStatus() {
     return result;
 }
 
-void SystemManager::LaunchTitle(u64 titleID) {
-    if (!OSIsTitleInstalled(titleID)) {
-        return;
+void RichPresenceManager::setGameState(std::string_view state, std::string_view details,
+        s64 startTimestamp) {
+    if (!m_statusWasSent || state != std::string_view(status().state) ||
+            startTimestamp != status().startTimestamp) {
```
Comment (riidefi) [inline payload/game/host_system/SystemManager.cc:343]: status().state != || status().details != details || status().startTimestamp  != startTimestamp

File: payload/game/host_system/SystemManager.cc [hunk 9]
```diff
@@ -115,11 +337,16 @@ bool RichPresenceManager::sendStatus() {
...
+        // TODO: replace m_state and details with BMG strings
+        status().state = state;
+        status().details = details;
+        status().startTimestamp = startTimestamp;
+        sendStatus();
```
Comment (riidefi) [inline payload/game/host_system/SystemManager.cc:348]:
```diff
status().details = details;
status().startTimestamp = startTimestamp;
+ m_statusWasSent = false;
sendStatus();
```

File: payload/game/host_system/SystemManager.cc [hunk 10]
```diff
@@ -103,9 +113,221 @@ void RichPresenceManager::terminateConnection() {
...
+            break;
+        default:
+            break;
+        }
+        break;
+    }
+
+    if (sceneId == RKSceneID::Race) {
+        auto timestamp = SP::IOSDolphin::GetSystemTime();
+        startTimestamp = timestamp ? *timestamp : 0;
```
[the remark this answers is not in the document]
Comment (vabold) [reply payload/game/host_system/SystemManager.cc]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/ce9a5b71f5ed43ce29942a42e7ee2572c975079a.

File: payload/game/host_system/SystemManager.cc (hunk 9 above)
Comment (vabold) [reply payload/game/host_system/SystemManager.cc:351]: Resolved in https://github.com/stblr/mkw-sp/pull/484/commits/ce9a5b71f5ed43ce29942a42e7ee2572c975079a.

File: payload/game/host_system/SystemManager.cc (hunk 8 above)
Comment (vabold) [reply payload/game/host_system/SystemManager.cc:345]: Potential resolution in https://github.com/stblr/mkw-sp/pull/484/commits/ce9a5b71f5ed43ce29942a42e7ee2572c975079a. The goal is to enforce the design principle of details being groups of states, where one state does not belong to multiple groups.

Comment (riidefi) [reply payload/game/host_system/SystemManager.cc:345]: I would prefer if `setGameState` makes no assumptions about the caller here.

**Review by stblr — APPROVED** [review]

**Pull request merged**